### PR TITLE
gsoc: Fix colors from the output

### PR DESCRIPTION
**Pull request by Nosferatul**

### CODE REVIEW

File: coalib/output/ConsoleInteraction.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ def highlight_text(no_color, text, lexer=TextLexer(), style=None):
 
 
 def color_letter(console_printer, line):
+    endc = '\033[0m'
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Can you make this variable a level higher? I don't like seeing it repeatedly all over the code :/

Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: i'll just remove them, it doesn't matter

File: coalib/output/ConsoleInteraction.py [hunk 2]
```diff
@@ -519,9 +520,10 @@ def print_affected_lines(console_printer, file_dict, sourcerange):
 
     if sourcerange.start.line is not None:
         if len(file_dict[sourcerange.file]) < sourcerange.end.line:
-            console_printer.print(format_lines(lines=STR_LINE_DOESNT_EXIST,
-                                               line_nr=sourcerange.end.line,
-                                               symbol='!'))
+            console_printer.print(format_lines(
```
Comment (adhikasp) [inline coalib/output/ConsoleInteraction.py]: What do you  change here?

Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: i think it was something, but i deleted it :smile:

**Comment by Nosferatul** [comment]
with this pr  http://imgur.com/80DEvWe. it's lxterminal

File: coalib/output/ConsoleInteraction.py [hunk 3]
```diff
@@ -255,7 +255,8 @@ def print_lines(console_printer,
             console_printer.print('')
         else:
             console_printer.print(highlight_text(
-                no_color, line[printed_chars:], lexer), end='')
+                no_color, line[printed_chars:], lexer, BackgroundMessageStyle),
+                                  end='')
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Wouldn't it be better if we add this to highlight_text itself?

Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Since logically we want to end the highlight at some point and not have it bled to the other lines.

Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: Hmm, I don't think we want that `style` variable from `highlight_text` be non-optional

Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: so, i think `style` variable from `highlight_text` should be non-optional, right?

Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Well, we wanna make sure by doing that doesn't break anything.

**Comment by yukiisbored** [comment]
Okay, this looks good, I'm gonna ack this and leave the merging to @Adrianzatreanu.

**Comment by yukiisbored** [comment]
Tested it locally, it looks good :)

**Comment by yukiisbored** [comment]
Can someone else try it on `gnome-terminal` and `konsole`? I can't test it because it requires a lot of requirements.

**Comment by yukiisbored** [comment]
@adtac Tested it on `st`

st

**Comment by yukiisbored** [comment]
@hemangsk @nkprince007 Can you test it on macOS?

**Comment by hemangsk** [comment]
on it

**Comment by hemangsk** [comment]
From Terminal.app : 

screen shot 2017-08-18 at 4 31 26 pm

**Comment by Adrianzatreanu** [comment]
https://ibb.co/b0dQsQ
how it looks on my mac

**Comment by yukiisbored** [comment]
@Adrianzatreanu euhm, that's coala master. This PR isn't merged yet.

**Comment by yukiisbored** [comment]
Okay, I'm not sure what's wrong with macOS Terminal.app because we've set the BackgroundMessageStyle after the text is highlighted.

**Comment by yukiisbored** [comment]
Okay, I just noticed @adtac screenshot is coala release

**Comment by yukiisbored** [comment]
Comparison between current master (2017-08-19) to this PR.

**Comment by yukiisbored** [comment]
Okay, now I'm confused. Isn't the code supposed to highlight the part of the code? Why isn't it highlighted?

File: coalib/output/ConsoleInteraction.py [hunk 4]
```diff
@@ -241,21 +238,24 @@ def print_lines(console_printer,
         printed_chars = 0
         if i == sourcerange.start.line and sourcerange.start.column:
             console_printer.print(highlight_text(
-                no_color, line[:sourcerange.start.column - 1], lexer), end='')
+                no_color, line[:sourcerange.start.column - 1],
+                BackgroundMessageStyle, lexer), end='')
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: I think this should be BackgroundMessageStyle?

File: coalib/output/ConsoleInteraction.py [hunk 5]
```diff
@@ -241,21 +238,24 @@ def print_lines(console_printer,
...
 
             printed_chars = sourcerange.start.column - 1
 
         if i == sourcerange.end.line and sourcerange.end.column:
             console_printer.print(highlight_text(
                 no_color, line[printed_chars:sourcerange.end.column - 1],
-                lexer, BackgroundSourceRangeStyle), end='')
+                BackgroundSourceRangeStyle, lexer), end='')
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: and this should be BackgroundSourceRangeStyle?

File: coalib/output/ConsoleInteraction.py [hunk 6]
```diff
@@ -241,21 +238,24 @@ def print_lines(console_printer,
...
 
             console_printer.print(highlight_text(
-               no_color, line[sourcerange.end.column - 1:], lexer), end='')
+               no_color, line[sourcerange.end.column - 1:],
+               BackgroundSourceRangeStyle, lexer), end='')
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Not sure about this one and 254, I think it's a hack to fix the color issue?

**Comment by yukiisbored** [comment]
@Nosferatul Can you euhm, redo that code which highlight parts of the code? I think it's need to be redone since it's pretty confusing and probably explains some of the hacks done to fix the colors.

**Comment by yukiisbored** [comment]
Is it because it requires 256 colors to be enabled?

**Comment by yukiisbored** [comment]
I think there's something wrong with this change that makes the code not highlighted.

**Comment by yukiisbored** [comment]
Okay, I tried playing with pygments in a Python REPL and I think `bg:` doesn't work.

**Comment by Nosferatul** [comment]
> Okay, now I'm confused. Isn't the code supposed to highlight the part of the code? Why isn't it highlighted?

I think it is from the screenshot and the problem with green color is because we don't have provide a style for lexer (that's what I understood when I tried it)

**Comment by adtac** [comment]
@yukiisbored my screenshot was after checking out @Nosferatul 's fork and it was in the `fix_color` branch.

**Comment by yukiisbored** [comment]
@adtac Euhm, I think you ran `coala` instead of `./coala`

**Comment by adtac** [comment]
`./coala` results:

**Comment by yukiisbored** [comment]
@Nosferatul Makes sense, since we're abusing a syntax highlighter here 😟

**Comment by yukiisbored** [comment]
Okay, I tested on macOS Sierra. It works. most likely because @hemangsk is running the `pre` version instead of this branch.

`iTerm`:
a

`Terminal.app`:
b

**Comment by yukiisbored** [comment]
@Nosferatul One more little change, the line number before the diff bugs me. Can you make it the same as the one the diff (the line number is inside the `[]`)?

**Comment by Nosferatul** [comment]
http://imgur.com/ke0Q0Mw

**Comment by yukiisbored** [comment]
ack https://github.com/coala/coala/pull/4627/commits/281a1096eedd58b1a2fbda428c2500e5794d53b9

**Comment by yukiisbored** [comment]
ack https://github.com/coala/coala/pull/4627/commits/5518d7b0027cdcf77813eeec69f9375bfc0c419d

**Comment by yukiisbored** [comment]
gonna leaver merge to @Adrianzatreanu so he can take another look

**Comment by Adrianzatreanu** [comment]
ack from me too, just rebase and merge it @Nosferatul

**Comment by Nosferatul** [comment]
ack 377585d

**Comment by Nosferatul** [comment]
reack 377585d

**Comment by Nosferatul** [comment]
reack d03303c

**Comment by Nosferatul** [comment]
@rultor merge

**Pull request merged**